questionpy_sdk/webserver/question_ui.py

```diff
@@ -169,37 +179,29 @@ def get_metadata(self) -> QuestionMetadata:
 
         return question_metadata
 
-    def render(self, attempt: dict | None = None, options: QuestionDisplayOptions | None = None) -> str:
-        """Applies transformations to the descendants of a given node and returns the resulting HTML."""
-        copy = deepcopy(self._root)
-        copy.nsmap.setdefault(None, self.XHTML_NAMESPACE)  # Interpret the default namespace as XHTML.
-        newdoc = etree.ElementTree(copy)
-
-        xpath = etree.XPathDocumentEvaluator(newdoc)
-        xpath.register_namespace("xhtml", self.XHTML_NAMESPACE)
-        xpath.register_namespace("qpy", self.QPY_NAMESPACE)
-
-        self._resolve_placeholders(xpath)
-        self._hide_unwanted_feedback(xpath, options)
-        self._hide_if_role(xpath, options)
-        self._set_input_values_and_readonly(xpath, attempt, options)
-        self._soften_validation(xpath)
-        self._defuse_buttons(xpath)
-        self._shuffle_contents(xpath)
-        self._add_styles(xpath)
-        self._format_floats(xpath)
+    def render(self) -> str:
```

Comment:
Ohne das Kopieren wird in dieser Funktion `self._xml` verändert, richtig? Bedeutet das nicht, dass `get_metadata` danach auch einen anderen Wert zurückgibt?

Reply:
Richtig, wir könnten noch dokumentieren, dass `get_metadata` vor `render` aufgerufen werden sollte, falls die Metadaten überhaupt gebraucht werden.

Die Metadaten werden ohnehin nur aus `formulation` extrahiert. Ich überlege gerade, ob wir noch eine Unterklasse `QuestionFormulationUIRenderer` erstellen sollten, die in `init` schon die Metadaten extrahiert.

Reply:
Vielleicht sollten wir so eine Unterklasse direkt für jede Art von UI erstellen (und die Basisklasse abstrakt machen), falls wir mal Anpassungen an der Darstellung einer bestimmten UI vornehmen möchten.

Reply:
Wenn man diese Funktion ein zweites Mal aufruft, sollte am besten direkt das fertig gerenderte HTML zurückgegeben werden

Reply:
Ja, gute Idee.